### PR TITLE
[https://nvbugs/6638489][fix] Robustness and functional enhancement of MNNVL Allreduce

### PR DESCRIPTION
**Pull request by timlee0212**
## Dev Engineer Review

- Strengthens MNNVL memory ownership, allocation alignment, capability checks, invariant validation, and rollback-safe cleanup.
- Replaces `is_mnnvl` heuristics with `_MnnvlDispatchPlan` and rank-consistent runtime qualification.
- Adds `MnnvlWorkspace` and `MnnvlTransport` bindings.
- Adds workspace reuse, locking, buffer validation, fallback handling, and CUDA Graph safeguards.
- Supports `AllReduceFusionOp.RMS_NORM` without residuals. The operation requires `gamma` and `epsilon`.
- Updates buffer sizing, pointer access, workspace access, and `noexcept` destruction APIs.
- No configuration or test-list changes were identified.

## QA Engineer Review

- Added `test_mnnvl_rms_norm_fusion`.
- Added `run_mnnvl_rms_norm_single_rank`.
- Added `RMS_NORM_SHAPE_CASES`.
- Covers one-shot, two-shot, and CGA hidden-size cases across MPI ranks.
- The tests are not listed in `tests/integration/test_lists/` based on the provided changes.
- Verdict: insufficient. Add the tests to the appropriate `test-db/` or `qa/` list, or provide equivalent CI coverage.

## Description

This PR includes several robustness enhancement to the MNNVL allreduce op and the underlying multicast buffer implementation.

### Robustness fix for mcastbuffer

- Exchange fabric handles through checked, owned storage so ranks no longer use double-scaled pointer arithmetic or aliased MPI send/receive buffers.
- Allocate one multicast-granularity-aligned buffer and expose only usable workspace capacity before the signal pad, so Python no longer needs a separate growth heuristic.
- Validate communicator, request, and granularity invariants, and set the GPUDirect-RDMA allocation flag only when the device reports that capability.
- Tear down unicast and multicast virtual-address resources in a non-throwing, rollback-safe order.
- Keep multicast allocation ownership in returned tensor views and synchronize the global multicast-pointer registry.
- Serialize Python workspace creation and replacement, reuse one workspace per mapping, and reject growth during CUDA Graph capture.

### Add RMS Norm without residual pattern support

- Add the missing support for the fusion pattern`AllReduceFusionOp.RMS_NORM` to the existing all-reduce + RMSNorm kernel path, requiring only `gamma` and `epsilon`.

### Revise dispatch check & error handling

- Replace the old `is_mnnvl` gate with a Python `_MnnvlDispatchPlan`: POSIX-FD for a single-node TP group, FABRIC for a cross-node group, and `TRTLLM_FORCE_MNNVL_AR_FABRIC=1` as the diagnostic override.
- Add a nanobind `MnnvlWorkspace` that owns runtime qualification and multicast allocation, exposing `is_available`, `reason`, usable size, and the local tensor view.
- Have C++ preflight the CUDA device, compute capability, host locality, multicast/fabric attributes, live fabric create/export/import, and unique physical GPUs, then agree the result across ranks.
- Treat expected unavailability as fallback (`is_available is False`) while still throwing unexpected errors; Python catches only `_MnnvlUnavailableError`.
- Warn when an explicit `MNNVL` strategy falls back to `AUTO`, and log the same reason at debug level for opportunistic `AUTO` probing.

---

## Test Coverage

`tests/unittest/_torch/multi_gpu/test_mnnvl_allreduce.py` and any multi-node model test with TP/TEP pattern.

## PR Checklist

Please review the following before submitting your PR:
- PR description clearly explains what and why. If using CodeRabbit's summary, please make sure it makes sense.
- PR Follows [TRT-LLM CODING GUIDELINES](https://github.com/NVIDIA/TensorRT-LLM/blob/main/CODING_GUIDELINES.md) to the best of your knowledge.
- Test cases are provided for new code paths (see [test instructions](https://github.com/NVIDIA/TensorRT-LLM/tree/main/tests#1-how-does-the-ci-work))
- If PR introduces API changes, an appropriate PR label is added - either `api-compatible` or `api-breaking`. For `api-breaking`, include `BREAKING` in the PR title.
- Any new dependencies have been scanned for license and vulnerabilities
- [CODEOWNERS](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/CODEOWNERS) updated if ownership changes
- Documentation updated as needed
- Update [tava architecture diagram](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/tava_architecture_diagram.md) if there is a significant design change in PR.
- The reviewers assigned automatically/manually are appropriate for the PR.

- [x] Please check this after reviewing the above items as appropriate for this PR.

## GitHub Bot Help

To see a list of available CI bot commands, please comment `/bot help`.

### CODE REVIEW

File: cpp/tensorrt_llm/thop/allreduceOp.cpp [hunk 1]
```diff
@@ -2095,12 +2097,10 @@ std::vector<torch::Tensor> mnnvlFusionAllReduce(torch::Tensor& input, torch::opt
 
     if (hasRmsNormFusion)
     {
-        TORCH_CHECK(residual_in.has_value() && gamma.has_value() && epsilon.has_value(),
-            "[mnnvlFusionAllReduce] residual_in, gamma, and epsilon must be provided for RMSNorm fusion");
-        TORCH_CHECK(residual_in.value().is_contiguous(), "[mnnvlFusionAllReduce] residual_in must be contiguous");
+        TORCH_CHECK(gamma.has_value() && epsilon.has_value(),
+            "[mnnvlFusionAllReduce] gamma and epsilon must be provided for RMSNorm fusion");
         TORCH_CHECK(gamma.value().is_contiguous(), "[mnnvlFusionAllReduce] gamma must be contiguous");
```
[the remark this answers is not in the document]
Comment (timlee0212) [reply cpp/tensorrt_llm/thop/allreduceOp.cpp:2103]: Waive this comment as we want to keep the C++ launch path clean and avoid unnecessary check. Python already supplies RMSNorm.weight as [hidden] on the same device/dtype as the activation, so these checks are redundant on the supported path.

File: tensorrt_llm/_torch/distributed/ops.py [hunk 2]
```diff
@@ -790,24 +946,29 @@ def __init__(self,
...
+                    self.mnnvl_allreduce = MNNVLAllReduce(
+                        self.mapping,
+                        dtype,
+                        dispatch_plan,
+                        warn_on_fallback=is_explicit_mnnvl,
+                    )
+                except _MnnvlUnavailableError as error:
+                    log_fallback(
+                        f"MNNVL AllReduce is unavailable: {error}. Falling back to AUTO."
                     )
```
[the remark this answers is not in the document]
Comment (timlee0212) [reply tensorrt_llm/_torch/distributed/ops.py:962]: Addressed in the latest commit, please check again

File: tensorrt_llm/_torch/distributed/ops.py [hunk 3]
```diff
@@ -555,111 +673,132 @@ class MNNVLAllReduce(nn.Module):
...
+        if input.ndim < 2 or input.shape[-1] <= 0:
+            logger.debug(
+                f"[MNNVL AllReduce] Input must have at least two dimensions and a positive hidden dimension, "
+                f"got shape {tuple(input.shape)}.")
+            return None
+
         fusion_op = all_reduce_params.fusion_op
+        is_fusion = fusion_op != AllReduceFusionOp.NONE
+        if is_fusion and fusion_op not in MNNVLAllReduce.SUPPORTED_FUSION_OPS:
+            logger.debug(
```
[the remark this answers is not in the document]
Comment (timlee0212) [reply tensorrt_llm/_torch/distributed/ops.py:765]: Again, this contract is fulfilled by the caller.
